### ISSUE TITLE
[psyir] Routine arguments given intent(inout) by default

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#1306]
Currently any dummy arguments within a routine that do not have an explicit intent are re-generated as having `intent(inout)`. If it turns out that these should actually have been `intent(in)` and there are other routines that pass in variables that *are* themselves declared as `intent(in)` then the compiler (gfortran) gets upset.

We could either:

 1. Not add an intent by default, or
 2. Use dependence analysis to work out what the intent should be if it's not specified

**Comment by arporter** [comment]
Given time pressure and robustness, I think 1. sounds like the best bet.

**Comment by sergisiso** [comment]
May be worth checking we are not losing any performance with this, as I think compilers might use the intent attributes to give some guarantees to the calling interface and optimize it (e.g. avoid extra tmp copies). I don't know if inout nad no-intent have different performance implications.

**Comment by arporter** [comment]
Yes, I don't know either and it's going to vary from compiler to compiler. For now I think I need to get the fix on ASAP but I will add a TODO that we should use dependence analysis to try to figure out what the intent should be. We should be able to do that unless the symbol is itself then passed as an argument to another Routine for which we don't have interface information available.

**Comment by arporter** [comment]
Happily, it is a tiny change to the frontend and the rest works as expected. The text for the TODO was longer :-)

**Comment by sergisiso** [comment]
Reviewing the PR this is less than an issue that I thought, I was under the impression we were dropping information but this is really that we don't have the intent information to begin with, so I don't mind the current code already be the good solution.
